### PR TITLE
Fix register fails on channles requests. #137

### PR DESCRIPTION
**Pull request by shmoon-kr**
Fixed register mutation not working in channels environment. Instead of the get_current_site() function, we implemented it by getting the required values from the request.headers object and the request.consumer.scope object and returning them. There are a few things that need to be reviewed.
1. the content of the request object has changed - this is included in the context used when rendering html mail. If they want to include the contents of the request in the email, they will need to customize it.
2. I've tested and found that the site_name and domain returned by get_mail_context() actually contain the same values, so it should work the same in a channels environment. However, if your site_name and domain are different, you may run into issues.
3. I used the port number by parsing request.header["host"]. if the request.header contains anything other than a valid "host:port" form of the value, an error may occur.

### CODE REVIEW

**Comment by gqlauth** [comment]
### Release Notes
<details>
<summary>Just wanted to say</summary>

success

</details>
Release type: patch

Fix register fails on channles requests. #137
Do noyt call get_current_site() from get_email_context()
Extract values from channels context.

File: gqlauth/models.py [hunk 1]
```diff
@@ -58,20 +58,35 @@ def send(self, subject, template, context, recipient_list=None):
...
-            "domain": site.domain,
-            "protocol": "https" if request.is_secure() else "http",
-            "path": path,
-            "timestamp": time.time(),
-            **app_settings.EMAIL_TEMPLATE_VARIABLES,
-        }
+        if isinstance(info.context, dict):
+            request = info.context["request"]
+            return {
+                "user": self.user,
```
Comment (nrbnlulu) [inline gqlauth/models.py:64]: try to do something like
```py
is_channels = isinstance(info.context, dict)

if is_channels:
	domain, port =  request.headers["host"].split(":")
else:
	domain = site.domain
	# yada yada
return {
	"domain" domain,
	# yada yada
```

**Review by nrbnlulu — COMMENTED** [review]
Also please add tests

File: gqlauth/models.py [hunk 2]
```diff
@@ -58,16 +58,28 @@ def send(self, subject, template, context, recipient_list=None):
...
+            domain, port = request.headers["host"].split(":")
+            site_name = domain
+            protocol = request.consumer.scope["type"]
+        else:
+            request = info.context.request
+            site = get_current_site(request)
+            port = request.get_port()
+            site_name = site.name
+            domain = site.domain
+            protocol = ("https" if request.is_secure() else "http",)
```
Comment (nrbnlulu) [inline gqlauth/models.py:73]: that's a tuple :upside_down_face:

**Comment by nrbnlulu** [comment]
> run the schema.execute() function in the channels context: I haven't seen this done, if you know how, please let me know.
You can try to use the tests consumer 
https://github.com/nrbnlulu/strawberry-django-auth/blob/710017565314d6403f1fadb79fd48cd6616ff044/tests/test_middlewares.py#L13

BTW TBH I am not sure that you should use channels for mutation / queries... Do you have any reason to do so?

File: gqlauth/models.py (hunk 2 above)
Comment (nrbnlulu) [reply gqlauth/models.py:73]: No it wasn't. but you what you wrote is a tuple
`("https" if request.is_secure() else "http",)`
`
```suggestion
            protocol = "https" if request.is_secure() else "http",
```

Comment (shmoon-kr) [reply gqlauth/models.py:73]: It seems to have been automatically fixed by [pre-commit-ci[bot]. I'd like you to take a look at this bot.

**Comment by shmoon-kr** [comment]
As per your suggestion, I wrote a test code to call register mutation in channels context using verified_user_status_type andverified_channels_app_communicator. However, it seems to fail to execute. When I ran the same code in sync context, it worked fine. Do you have any idea how to fix it?

**Comment by shmoon-kr** [comment]
The problem is that it's trying to register a new user, and the username and phone_number already exist. I fixed it and republished the PR, so please review it.

File: tests/test_register.py [hunk 3]
```diff
@@ -50,6 +50,17 @@ def test_register_invalid_password_validation(verified_user_status_type, anonymo
     assert executed["errors"]
 
 
+async def test_channels_register(
+    verified_user_status_type, captcha, verified_channels_app_communicator
```
Comment (nrbnlulu) [inline tests/test_register.py:54]: You don't need a verified consumer here I think

**Review by nrbnlulu — APPROVED** [review]
Generally I accept the changes though I want to test this manually because the consumer is having problems ATM.

File: tests/test_register.py (hunk 3 above)
Comment (shmoon-kr) [reply tests/test_register.py]: You're right, I replaced it with unverified_channels_app_communicator, as it doesn't require authentication when registering.

**Pull request merged**